storage/grpc_metrics.go

```diff
@@ -0,0 +1,295 @@
+	boundary := 0.0
+	increment := 0.002
+	// 2ms buckets for first 100ms, so we can have higher resolution for uploads and downloads in the 100 KiB range
+	for i := 0; i < 50; i++ {
+		boundaries = append(boundaries, boundary)
+		// increment by 2ms
+		boundary += increment
+	}
+	// For the remaining buckets do 10 10ms, 10 20ms, and so on, up until 5 minutes
+	for i := 0; i < 150 && boundary < 300; i++ {
+		boundaries = append(boundaries, boundary)
+		if i != 0 && i%10 == 0 {
+			increment *= 2
+		}
+		boundary += increment
+	}
+	return boundaries
+}
+
+func sizeHistogramBoundaries() []float64 {
+	kb := 1024.0
+	mb := 1024.0 * kb
+	gb := 1024.0 * mb
+	boundaries := []float64{}
+	boundary := 0.0
+	increment := 128 * kb
+	// 128 KiB increments up to 4MiB, then exponential growth
+	for len(boundaries) < 200 && boundary <= 16*gb {
+		boundaries = append(boundaries, boundary)
+		boundary += increment
+		if boundary >= 4*mb {
+			increment *= 2
+		}
+	}
+	return boundaries
+}
+
+func metricFormatter(m metricdata.Metrics) string {
+	return "storage.googleapis.com/client/" + strings.ReplaceAll(string(m.Name), ".", "/")
+}
+
+func gcpAttributeExpectedDefaults() []attribute.KeyValue {
+	return []attribute.KeyValue{
+		{Key: "location", Value: attribute.StringValue("global")},
+		{Key: "cloud_platform", Value: attribute.StringValue("unknown")},
+		{Key: "host_id", Value: attribute.StringValue("unknown")}}
+}
+
+// Added to help with tests
+type internalPreparedResource struct {
+	projectToUse string
+	resource     *resource.Resource
+}
+
+func newPreparedResource(ctx context.Context, project string, resourceOptions []resource.Option) (*internalPreparedResource, error) {
+	detectedAttrs, err := resource.New(ctx, resourceOptions...)
+	if err != nil {
+		return nil, err
+	}
+	preparedResource := &internalPreparedResource{}
+	s := detectedAttrs.Set()
+	p, present := s.Value("cloud.account.id")
+	if present {
+		preparedResource.projectToUse = p.AsString()
+	} else {
+		preparedResource.projectToUse = project
+	}
+	updates := []attribute.KeyValue{}
+	for _, kv := range gcpAttributeExpectedDefaults() {
+		if val, present := s.Value(kv.Key); !present || val.AsString() == "" {
+			updates = append(updates, attribute.KeyValue{Key: kv.Key, Value: kv.Value})
+		}
+	}
+	r, err := resource.New(
+		ctx,
+		resource.WithAttributes(
+			attribute.KeyValue{Key: "gcp.resource_type", Value: attribute.StringValue(monitoredResourceName)},
+			attribute.KeyValue{Key: "instance_id", Value: attribute.StringValue(uuid.New().String())},
+			attribute.KeyValue{Key: "project_id", Value: attribute.StringValue(project)},
+			attribute.KeyValue{Key: "api", Value: attribute.StringValue("grpc")},
+		),
+		resource.WithAttributes(detectedAttrs.Attributes()...),
+		// Last duplicate key / value wins
+		resource.WithAttributes(updates...),
+	)
+	if err != nil {
+		return nil, err
+	}
+	preparedResource.resource = r
+	return preparedResource, nil
+}
+
+type internalMetricsConfig struct {
```

Comment:
I don't think this needs to be a separate type, IIUC it's just being used to pass around these two values correct?